lntest/itest/lnd_coopclose_test.go

```diff
@@ -0,0 +1,148 @@
+	)
+
+	// Create a hodl invoice for Bob. This will allow us to exercise the
+	// Shutdown logic.
+	var (
+		preimage = lntypes.Preimage{1, 2, 3}
+		payHash  = preimage.Hash()
+	)
+
+	invoiceReq := &invoicesrpc.AddHoldInvoiceRequest{
+		Value:      30000,
+		CltvExpiry: 40,
+		Hash:       payHash[:],
+	}
+
+	ctxt, cancel := context.WithTimeout(ctxb, defaultTimeout)
+	defer cancel()
+	bobInvoice, err := net.Bob.AddHoldInvoice(ctxt, invoiceReq)
+	require.NoError(t.t, err)
+
+	// Alice will now pay this invoice and get a single HTLC locked-in on
+	// each of the commitment transactions.
+	_, err = net.Alice.RouterClient.SendPaymentV2(
+		ctxb, &routerrpc.SendPaymentRequest{
+			PaymentRequest: bobInvoice.PaymentRequest,
+			TimeoutSeconds: 60,
+			FeeLimitMsat:   noFeeLimitMsat,
+		},
+	)
+	require.NoError(t.t, err)
+
+	waitForInvoiceAccepted(t, net.Bob, payHash)
+
+	// Alice and Bob should both have a single HTLC locked in.
+	nodes := []*lntest.HarnessNode{net.Alice, net.Bob}
+	err = wait.NoError(func() error {
+		return assertActiveHtlcs(nodes, payHash[:])
+	}, defaultTimeout)
+	require.NoError(t.t, err)
+
+	closeReq := &lnrpc.CloseChannelRequest{
+		ChannelPoint: chanPoint,
+	}
+
+	// Alice will initiate the coop close and we'll obtain the close
+	// channel client to assert that the channel is closed once the htlc is
+	// settled.
+	ctxt, cancel = context.WithTimeout(ctxb, defaultTimeout)
+	defer cancel()
+	closeRespStream, err := net.Alice.CloseChannel(ctxt, closeReq)
+	require.NoError(t.t, err)
+
+	time.Sleep(time.Second * 10)
+
+	// Bob will now settle the invoice.
+	settle := &invoicesrpc.SettleInvoiceMsg{
+		Preimage: preimage[:],
+	}
+	_, err = net.Bob.SettleInvoice(ctxt, settle)
+	require.NoError(t.t, err)
+
+	// The coop close negotiation should now complete.
+	var closeTxid *chainhash.Hash
+	err = wait.NoError(func() error {
+		closeResp, err := closeRespStream.Recv()
+		if err != nil {
+			return fmt.Errorf("unable to Recv() from stream: %w",
+				err)
+		}
+
+		pendingClose, ok := closeResp.Update.(*lnrpc.CloseStatusUpdate_ClosePending)
+		if !ok {
+			return fmt.Errorf("expected channel close update, "+
+				"instead got %v", pendingClose)
+		}
+
+		closeTxid, err = chainhash.NewHash(
+			pendingClose.ClosePending.Txid,
+		)
+		if err != nil {
+			return fmt.Errorf("unable to decode closeTxid: %w",
+				err)
+		}
+
+		coopTx, err := waitForTxInMempool(
+			net.Miner.Client, minerMempoolTimeout,
+		)
+		if err != nil {
+			return err
+		}
+
+		if !coopTx.IsEqual(closeTxid) {
+			return fmt.Errorf("mempool tx is not coop close tx")
+		}
+
+		return nil
+	}, channelCloseTimeout)
+	require.NoError(t.t, err)
+
+	// The closing transaction will be mined and Alice and Bob should both
+	// see the channel as successfully resolved.
+	block := mineBlocks(t, net, 1, 1)[0]
+	assertTxInBlock(t, block, closeTxid)
+
+	// Sleep again to ensure the ChannelArbitrator can advance to
+	// StateFullyResolved.
+	time.Sleep(10 * time.Second)
```

Comment:
Use a wait / predicate until the closed channel shows up in Alice + Bob's close channel lists?

Reply:
I added a predicate here, but for the other sleep I'm not sure it's possible. We need to start the coop close flow without settling too early because we want to test the behavior of trying to coop close while the HTLC is in-flight. I'm not aware of any RPC signals we can use to determine when `Shutdown` has been sent. If there were, we'd wait to send and receive `Shutdown`, and then call `SettleInvoice`.